crates/ra_assists/src/handlers/add_function.rs

```diff
@@ -0,0 +1,338 @@
+use ra_syntax::{
+    ast::{self, AstNode},
+    SmolStr, SyntaxKind, SyntaxNode, TextUnit,
+};
+
+use crate::{Assist, AssistCtx, AssistId};
+use ast::{CallExpr, Expr};
+use hir::ModuleDef;
+use ra_fmt::leading_indent;
+
+// Assist: add_function
+//
+// Adds a stub function with a signature matching the function under the cursor.
+//
+// ```
+// fn foo() {
+//      bar<|>("", baz());
+// }
+//
+// ```
+// ->
+// ```
+// fn foo() {
+//      bar("", baz());
+// }
+//
+// fn bar(arg_1: &str, baz: Baz) {
+//      todo!();
+// }
+//
+// ```
+pub(crate) fn add_function(ctx: AssistCtx) -> Option<Assist> {
+    let path: ast::Path = ctx.find_node_at_offset()?;
+
+    if ctx.sema.resolve_path(&path).is_some() {
+        // The function call already resolves, no need to add a function
+        return None;
+    }
+
+    if path.qualifier().is_some() {
+        return None;
+    }
+
+    let call: ast::CallExpr = ctx.find_node_at_offset()?;
+
+    let (generated_fn, cursor_pos, text_start) = generate_fn(&ctx, &call)?;
+
+    ctx.add_assist(AssistId("add_function"), "Add function", |edit| {
+        edit.target(call.syntax().text_range());
+
+        edit.set_cursor(cursor_pos);
+        edit.insert(text_start, generated_fn);
+    })
+}
+
+/// Generates a function definition that will allow `call` to compile.
+/// The function name must match.
+/// The arguments must have valid, nonconflicting names.
+/// The arguments' types must match those being passed to `call`.
+/// TODO generics?
+fn generate_fn(ctx: &AssistCtx, call: &CallExpr) -> Option<(String, TextUnit, TextUnit)> {
+    let (start, indent) = next_space_for_fn(&call)?;
+    let indent = if let Some(i) = &indent { i.as_str() } else { "" };
+    let mut fn_buf = String::with_capacity(128);
+
+    fn_buf.push_str("\n\n");
+    fn_buf.push_str(indent);
+    fn_buf.push_str("fn ");
+
+    let fn_name = fn_name(&call)?;
+    fn_buf.push_str(&fn_name);
+
+    let fn_generics = fn_generics(&call)?;
+    fn_buf.push_str(&fn_generics);
+
+    let fn_args = fn_args(ctx, &call)?;
+    fn_buf.push_str(&fn_args);
+
+    fn_buf.push_str(" {\n");
+    fn_buf.push_str(indent);
+    fn_buf.push_str("    ");
+
+    // We take the offset here to put the cursor in front of the `todo` body
+    let offset = TextUnit::of_str(&fn_buf);
+
+    fn_buf.push_str("todo!()\n");
+    fn_buf.push_str(indent);
+    fn_buf.push_str("}");
+
+    let cursor_pos = start + offset;
+    Some((fn_buf, cursor_pos, start))
+}
+
+fn fn_name(call: &CallExpr) -> Option<String> {
+    Some(call.expr()?.syntax().to_string())
+}
+
+fn fn_generics(_call: &CallExpr) -> Option<String> {
+    // TODO
+    Some("".into())
+}
+
+fn fn_args(ctx: &AssistCtx, _call: &CallExpr) -> Option<String> {
+    let arg_list = _call.syntax().descendants().find_map(ast::ArgList::cast)?;
+    for arg in arg_list.args() {
+        match &arg {
+            Expr::CallExpr(call_expr) => {
+                arg_for_call(ctx, call_expr);
```

Comment:
I'm a bit stumped as to how I can print the type of the function call arguments in the generated function.
Is this the correct approach?

Reply:
I think you want smth like `call.arg_list().map(|it| sema.type_of_expr(it))`.